clap-v3-utils/src/input_parsers/signer.rs

```diff
@@ -15,6 +18,90 @@ use {
 // Sentinel value used to indicate to write to screen instead of file
 pub const STDOUT_OUTFILE_TOKEN: &str = "-";
 
+#[derive(Debug)]
+pub struct SignerSourceParserBuilder {
+    prompt: bool,
+    file_path: bool,
+    usb: bool,
+    stdin: bool,
+    pubkey: bool,
+    allow_legacy: bool,
+}
+impl Default for SignerSourceParserBuilder {
+    fn default() -> Self {
+        Self {
+            prompt: true,
+            file_path: true,
+            usb: true,
+            stdin: true,
+            pubkey: true,
+            allow_legacy: true,
+        }
+    }
+}
+
+impl SignerSourceParserBuilder {
+    pub fn new() -> Self {
+        Self {
+            prompt: false,
+            file_path: false,
+            usb: false,
+            stdin: false,
+            pubkey: false,
+            allow_legacy: true,
+        }
+    }
+
+    pub fn allow_prompt(mut self) -> Self {
+        self.prompt = true;
+        self
+    }
+
+    pub fn allow_file_path(mut self) -> Self {
+        self.file_path = true;
+        self
+    }
+
+    pub fn allow_usb(mut self) -> Self {
+        self.usb = true;
+        self
+    }
+
+    pub fn allow_stdin(mut self) -> Self {
+        self.stdin = true;
+        self
+    }
+
+    pub fn allow_pubkey(mut self) -> Self {
+        self.pubkey = true;
+        self
+    }
+
+    pub fn disallow_legacy(mut self) -> Self {
+        self.allow_legacy = false;
+        self
+    }
+
+    pub fn build(self) -> ValueParser {
+        ValueParser::from(
+            move |arg: &str| -> Result<SignerSource, SignerSourceError> {
+                let signer_source = parse_signer_source(arg)?;
+                if !self.allow_legacy && signer_source.legacy {
+                    return Err(SignerSourceError::UnsupportedSource);
+                }
+                match signer_source.kind {
+                    SignerSourceKind::Prompt if self.prompt => Ok(signer_source),
+                    SignerSourceKind::Filepath(_) if self.file_path => Ok(signer_source),
+                    SignerSourceKind::Usb(_) if self.usb => Ok(signer_source),
+                    SignerSourceKind::Stdin if self.stdin => Ok(signer_source),
+                    SignerSourceKind::Pubkey(_) if self.pubkey => Ok(signer_source),
+                    _ => Err(SignerSourceError::UnsupportedSource),
```

Comment:
seems like we should be offering more guidance as to what makes a given signer unsupported.  we have the state to do so after all.  otherwise this could all just be a bunch of free getters (i don't see where any mixed use exists or is tested here?)

Reply:
Yeah this is a valid point. The main purpose of `SignerSourceParserBuilder` was to really replace the individual validation functions that exist in `input_validators.rs` in a modular/uniform way.

In `input_validators.rs` there are functions like `is_pubkey_or_keypair`, `is_keypair_or_ask_keyword`, `is_prompt_signer_source`, etc. If I look at the existing solana cli applications, the use of these validators do seem a little arbitrary at least at first sight (e.g. why not accept any signer or keypair source?) and there are not much documentation.

I think it could be best to add guiding docs to `SignerSourceKind` (since it will become a public struct in the follow-up PR), explaining each of the variants and when different variants are recommended to be used.

But perhaps this change could be done in a follow-up PR after the rest of the other changes are in from https://github.com/solana-labs/solana/pull/33478?